core/src/storage/lazy_chunk.rs

```diff
@@ -0,0 +1,393 @@
+                Push::push(value, key_ptr);
+            }
+            None => {
+                // Clear the storage at the given index.
+                crate::env::clear_contract_storage(key_ptr.next_for::<T>());
+            }
+        }
+    }
+}
+
+impl<T> Entry<T> {
+    /// Returns `true` if the cached value of the entry has potentially been mutated.
+    pub fn mutated(&self) -> bool {
+        self.mutated.get()
+    }
+
+    /// Returns a shared reference to the value of the entry.
+    pub fn value(&self) -> Option<&T> {
+        self.value.as_ref()
+    }
+
+    /// Returns an exclusive reference to the value of the entry.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry if the entry was occupied
+    /// since the caller could potentially change the returned value.
+    pub fn value_mut(&mut self) -> Option<&mut T> {
+        self.mutated.set(self.value.is_some());
+        self.value.as_mut()
+    }
+
+    /// Takes the value from the entry and returns it.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry if the entry was occupied.
+    pub fn take_value(&mut self) -> Option<T> {
+        self.mutated.set(self.value.is_some());
+        self.value.take()
+    }
+
+    /// Puts the new value into the entry and returns the old value.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry to `true` as long as at
+    /// least one of `old_value` and `new_value` is `Some`.
+    pub fn put(&mut self, new_value: Option<T>) -> Option<T> {
+        match new_value {
+            Some(new_value) => {
+                self.mutated.set(true);
+                self.value.replace(new_value)
+            }
+            None => self.take_value(),
+        }
+    }
+}
+
+impl<T> LazyChunk<T> {
+    /// Creates a new empty lazy chunk that cannot be mutated.
+    pub fn new() -> Self {
+        Self {
+            key: None,
+            cached_entries: UnsafeCell::new(EntryMap::new()),
+        }
+    }
+
+    /// Performs the given closure on the mutable cached entries.
+    ///
+    /// # Note
+    ///
+    /// Actions on the mutable lazy entries are performed within the closure
+    /// to not leak exclusive references to it to the outside. This is important
+    /// since the `for_entries` method itself operates only on `&self`.
+    fn for_cached_entries<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce(&mut EntryMap<T>) -> R,
+    {
+        // SAFETY: We operate on an exclusive reference on `BTreeMap` within the
+        //         given closure while our method receiver is only a shared reference.
+        //         However, due to encapsulation of the exclusive reference within
+        //         the given closure we cannot leak the exclusive reference outside
+        //         of the closure. So the below action is safe in this regard.
+        f(unsafe { &mut *self.cached_entries.get() })
```

Comment:
I'm really not sure that this is actually safe.
I don't know why, but it's very suspicious that `Cell` doesn't provide a API that provides a closure taking a mutable reference to its content, and indeed it seems that it's wrong to do so: https://github.com/rust-lang/rust/issues/50186#issuecomment-411169568

Reply:
It is very likely however that this isn't safe only because it's possible to do something very stupid that no-one would reasonably do.

Reply:
The helper function is completely broke: https://github.com/paritytech/ink/pull/311#discussion_r375966554
I will just remove it. Thanks for the pointers.